src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs

```diff
@@ -78,19 +71,25 @@ private long BinarySearchEstimate(long leftBound, long rightBound, long cap, Tra
                 }
             }
 
-            if (rightBound == cap && !TryExecutableTransaction(tx, header, rightBound))
+            if (rightBound == cap && !TryExecutableTransaction(tx, header, rightBound, token))
             {
                 return 0;
             }
 
             return rightBound;
         }
 
-        private bool TryExecutableTransaction(Transaction transaction, BlockHeader block, long gasLimit)
+        private bool TryExecutableTransaction(Transaction transaction, BlockHeader block, long gasLimit, CancellationToken token)
         {
             OutOfGasTracer tracer = new();
+
+            // TODO: Workaround to not mutate the original Tx
+            long originalGasLimit = transaction.GasLimit;
+
             transaction.GasLimit = gasLimit;
-            _transactionProcessor.CallAndRestore(transaction, block, tracer);
+            _transactionProcessor.CallAndRestore(transaction, block, tracer.WithCancellation(token));
+
+            transaction.GasLimit = originalGasLimit;
```

Comment:
We want to preserve the original Gas limit so we are forced to do this workaround. Ideally, `Transaction` would implement `ICloneable`, so we could make a copy with the desired amount of Gas instead of mutating the original one.

Reply:
IMO it is fine